content/post/project-dynamic-observation-and-interruption-of-sparql-queries.md

```diff
@@ -0,0 +1,313 @@
+
+QLever UI is a convenient web UI for the QLever backend. It makes it easy to use QLever from your browser. It also already
+rendered this tree in a human-readable form so everyone can easily analyse the bottlenecks of individual queries. Needless to
+say it wasn't too difficult to expand this functionality to update this visual representation based on messages coming
+from a newly created WebSocket connection instead of just a plain HTTP response. Of course, there are some pitfalls that
+need to be avoided, for example making sure the final HTTP response doesn't get replaced with a delayed WebSocket message.
+But apart from that this process was rather straightforward.
+
+![Example of a `RuntimeInformation` tree](/img/project-dynamic-observation-and-interruption-of-sparql-queries/RuntimeInformation%20Tree.png)
+
+<p style="text-align: center; font-size: 0.75em;">Figure 3: Example of how QLever UI would represent the runtime information tree.</p>
+
+#### `Boost.Beast`, `Boost.Asio` and Concurrent connections
+
+To implement all of this on the backend, the `Boost.Beast` library, built on top of `Boost.Asio` is used. It already
+served as the HTTP backbone for QLever for quite a while now. To be honest, I might be spoiled by HTTP server frameworks
+in a lot of other languages that do almost everything for you out of the box, but I always felt like even though `Boost.Beast`
+comes with a lot of ready-to-use components it still expects you to build the server on your own. It gives you a toolbox
+with all the parts to build an engine when in reality I would've liked to order the entire pre-built engine in the first place.
+I don't think I could build a better engine than experts on the matter, even if it isn't tailored to fit my exact use case.
+
+Because of this exposing a WebSocket endpoint in the backend was quite a hassle. The concept in itself wasn't really complex.
+When a new connection is being established and the query ID is not yet in use (usually this means the query has not started yet)
+the server just keeps the connection alive without sending anything else until a query with this specific ID starts. There can
+be multiple WebSockets waiting for the exact same query which doesn't really help keeping things simple, mainly to allow
+observations by server administrators in case queries are hogging up a lot of resources for no apparent reason. Once all
+listeners are connected to QLever and a query makes progress it has to broadcast the latest update to all listeners. Of course,
+there are some important considerations to be made here. The broadcast operation cannot be blocking in any way, otherwise,
+a slow WebSocket client can just slow down queries artificially. Also ideally all WebSockets should be able to operate
+independently of each other for maximum efficiency, but due to technical limitations of `Boost.Asio` we ended up running all
+io operations in a somewhat synchronized manner to avoid race conditions.
```

Comment:
I wouldn't say that this is true. Only very cheap ramping / async initiate functions are synchronized and that's the way it should be. Not need to pessimize your achievements here.